rubin_scheduler/scheduler/surveys/scripted_surveys.py

```diff
@@ -267,7 +272,28 @@ def set_script(self, obs_wanted):
         self.scheduled_obs = self.obs_wanted["mjd"]
 
     def generate_observations_rough(self, conditions):
-        observations = self._check_list(conditions)
-        observations = [self._slice2obs(obs) for obs in observations]
+        # if we have already called for this mjd, no need to repeat.
+        if self.last_mjd == conditions.mjd:
+            return self.observations
 
-        return observations
+        observations = self._check_list(conditions)
+        if observations is None:
+            self.observations = []
+            self.last_mjd = conditions.mjd
+            return self.observations
+
+        # If we want to ensure the observations can be completed before twilight starts
+        if self.before_twi_check:
+            # not bothering to include filter changes and read time now.
```

Comment:
Maybe you should include them, though? Particularly for ToO followups (which might be implemented with scripted surveys), they might ask for lots of filter changes, which could be significant.